File: li5/VideoViewController.m
//
//  VideoViewController.m
//  li5
//
//  Created by Martin Cocaro on 4/26/16.
//  Copyright © 2016 Li5, Inc. All rights reserved.
//

@import Li5Api;

#import "TeaserViewController.h"
#import "UnlockedViewController.h"
#import "VideoViewController.h"
#import "Li5SDK/Li5SDK-Swift.h"

@interface VideoViewController ()
{
    TeaserViewController *teaserViewController;
    UnlockedViewController *unlockedViewController;
    UIViewController<DisplayableProtocol> *currentViewController;
}

@end

@implementation VideoViewController

@synthesize product;

- (id)initWithProduct:(Product *)thisProduct andContext:(ProductContext)ctx
{
    self = [super init];
    if (self)
    {
        DDLogVerbose(@"");
        self.product = thisProduct;
        teaserViewController = [TeaserViewController teaserWithProduct:self.product andContext:ctx];
//        if (self.product.videoURL != nil && self.product.videoURL.length > 0)
//        {
//            unlockedViewController = [UnlockedViewController unlockedWithProduct:self.product andContext:ctx];
//        }
        currentViewController = nil;
    }
    return self;
}

- (void)viewDidLoad
{
    DDLogVerbose(@"");
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.automaticallyAdjustsScrollViewInsets = NO;
    [self.view setBackgroundColor: [UIColor redColor]];
    
    [self showViewController:teaserViewController withAppearanceTransition:NO];
}

- (void)viewWillAppear:(BOOL)animated
{
    DDLogVerbose(@"");
    [super viewWillAppear:animated];
    
    [currentViewController beginAppearanceTransition:YES animated:animated];
}

- (void)viewDidAppear:(BOOL)animated
{
    DDLogVerbose(@"");
    [super viewDidAppear:animated];
    
    [currentViewController endAppearanceTransition];
}

- (void)viewWillDisappear:(BOOL)animated
{
    DDLogVerbose(@"");
    [super viewWillDisappear:animated];
    
    [currentViewController beginAppearanceTransition:NO animated:animated];
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    DDLogVerbose(@"");
    
    [currentViewController endAppearanceTransition];
}

// From the container view controller
- (BOOL) shouldAutomaticallyForwardAppearanceMethods
{
    return YES;
}

- (void)hideViewController:(UIViewController<DisplayableProtocol> *)vc withAppearanceTransition:(BOOL)appear force:(BOOL)force
{
    DDLogVerbose(@"");
    [vc willMoveToParentViewController:nil];
    if (appear) [vc beginAppearanceTransition:NO animated:NO];
    if (force) [vc.view removeFromSuperview]; else [vc.view setAlpha:0.5];
    if (appear) [vc endAppearanceTransition];
    if (force) [vc removeFromParentViewController];
    [vc didMoveToParentViewController:nil];
}

- (void)showViewController:(UIViewController<DisplayableProtocol> *)vc withAppearanceTransition:(BOOL)appear
{
    DDLogVerbose(@"");
    [vc willMoveToParentViewController:self];
    [self addChildViewController:vc];
    vc.view.frame = self.view.bounds;
    [vc.view setAlpha:1.0];
    if (appear) [vc beginAppearanceTransition:YES animated:NO];
    [self.view addSubview:vc.view];
    if (appear) [vc endAppearanceTransition];
    [vc didMoveToParentViewController:self];
    currentViewController = vc;
}

- (BOOL)prefersStatusBarHidden
{
    return YES;
}

#pragma mark - User Actions

- (void)setPriority:(BCPriority)priority
{
    [teaserViewController setPriority:priority];
}

- (void)handleLongTap:(UITapGestureRecognizer *)sender
{
    DDLogVerbose(@"");
    unlockedViewController.initialPoint = [sender locationInView:teaserViewController.view];
    
    unlockedViewController.providesPresentationContextTransitionStyle = true;
    unlockedViewController.definesPresentationContext = true;
    unlockedViewController.modalPresentationStyle = UIModalPresentationOverFullScreen;
    
    [self showViewController:unlockedViewController withAppearanceTransition:YES];
    [self hideViewController:teaserViewController withAppearanceTransition:YES force:NO];
}

- (void)handleLockTap:(UIGestureRecognizer *)recognizer
{
    DDLogVerbose(@"Handling Lock Tap");
    
    [self hideViewController:unlockedViewController withAppearanceTransition:YES force:YES];
    [self showViewController:teaserViewController withAppearanceTransition:YES];
}

-(BCPlayer *)getPlayer {
    
    return [teaserViewController getPlayer];
}

#pragma mark - iOS Actions


//- (void)viewWillLayoutSubviews {
//    [super viewWillLayoutSubviews];
//    
//    teaserViewController.view.frame = self.view.bounds;
//    //    self.posterImageView.frame = self.view.bounds;
//}

-(void)dealloc
{
    DDLogDebug(@"%p",self);
    teaserViewController = nil;
    unlockedViewController = nil;
    currentViewController = nil;
}

- (void)didReceiveMemoryWarning
{
    DDLogDebug(@"");
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: Li5SDK/VideoViewController.swift
//
//  VideoViewController.swift
//  li5
//
//  Created by Sergio Daniel Lozano on 6/8/17.
//  Copyright © 2017 Li5, Inc. All rights reserved.
//

import UIKit


protocol VideoViewControllerProtocol : Displayable, PageIndexedProtocol {
    
}


class VideoViewController : UIViewController, VideoViewControllerProtocol, UIGestureRecognizerDelegate {
    
    // MARK: - PUBLIC INTERFACE
    
    // MARK: - Stored Properties
    
    var product: Product
    
    var teaserViewController: TeaserViewController!
    //TODO: var unlockedViewController: UnlockedViewController?
    
    var currentViewController: UIViewController?
    
    // MARK: - Computed Properties
    
    var pageIndex: Int
    
    // MARK: - Initializers
    
    
    required init(product: Product, context: PContext, pageIndex: Int) {
        self.product = product
        self.pageIndex = pageIndex
        
        teaserViewController = TeaserViewController.instance(withProduct: self.product,
                                                             andContext: context)
        
        if product.videoURL != nil && !product.videoURL.isEmpty {
            //  unlockedViewController = [UnlockedViewController unlockedWithProduct:self.product andContext:ctx];
        }
        
        currentViewController = nil
        
        super.init(nibName: nil, bundle: nil)
    }
    
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    // MARK: - Routines
    
    
    // MARK: - Lifecycle
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        automaticallyAdjustsScrollViewInsets = false
        view.backgroundColor = UIColor.clear
        
        present(viewController: teaserViewController, withAppearanceTransition: false)
    }
    
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        currentViewController?.beginAppearanceTransition(true, animated: animated)
    }
    
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        currentViewController?.endAppearanceTransition()
    }
    
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        currentViewController?.beginAppearanceTransition(false, animated: animated)
    }
    
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        currentViewController?.endAppearanceTransition()
    }
    
    
    // From the container view controller (???)
    override var shouldAutomaticallyForwardAppearanceMethods: Bool {
        return true
    }
    
    
    override var prefersStatusBarHidden: Bool {
        return true
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    
    deinit {
        teaserViewController = nil
        //TODO: unlockedViewController = nil
        currentViewController = nil
    }
    
    // MARK: - PRIVATE INTERFACE
    
    /// Present the given view controller calling the appearance methods and keeping the reference
    /// into the currentViewController variable
    /// - Parameters:
    ///   - vc: The viewController willing to present
    ///   - appearance: Whether it's appearing or not
    func present(viewController vc: UIViewController, withAppearanceTransition appearance: Bool) {
        log.verbose("Presenting VC from VideoViewController \(product.id)")
        
        vc.willMove(toParentViewController: self)
        addChildViewController(vc)
        vc.view.frame = view.bounds
        view.alpha = 1.0
        
        if appearance == true {
            vc.beginAppearanceTransition(true, animated: false)
        }
        
        view.addSubview(vc.view)
        
        if appearance == true {
            vc.endAppearanceTransition()
        }
        
        vc.didMove(toParentViewController: self)
        
        currentViewController = vc
    }
    
    
    /// Unpresents (dismisses) the given view controller calling the appearance methods and
    /// forcing its complete disappearance in case a true force flag is passed
    /// - Parameters:
    ///   - vc: The viewController willing to unpresent
    ///   - appearance: Whether it's appearing or not (???)
    ///   - force: Whether we need to force its complete disappearance or not
    func unpresent(viewController vc: UIViewController,
                   withAppearanceTransition appearance: Bool,
                   byForcing force: Bool) {
        
        log.verbose("Unpresenting VC from VideoViewController \(product.id)")
        
        vc.willMove(toParentViewController: nil)
        
        if appearance == true {
            vc.beginAppearanceTransition(false, animated: false)
        }
        
        if force {
            vc.view.removeFromSuperview()
        } else {
            vc.view.alpha = 0.5
        }
        
        if appearance == true {
            vc.endAppearanceTransition()
        }
        
        if force {
            vc.removeFromParentViewController()
        }
        
        vc.didMove(toParentViewController: nil)
    }
    
    
    func userDidLongTap(_ recognizer: UITapGestureRecognizer) {
        log.verbose("User did long tap")
        
//        unlockedViewController.initialPoint = [sender locationInView:teaserViewController.view];
//        
//        unlockedViewController.providesPresentationContextTransitionStyle = true;
//        unlockedViewController.definesPresentationContext = true;
//        unlockedViewController.modalPresentationStyle = UIModalPresentationOverFullScreen;
//        
//        [self showViewController:unlockedViewController withAppearanceTransition:YES];
//        [self hideViewController:teaserViewController withAppearanceTransition:YES force:NO];
    }
    
    
    func userDidTapForLock(_ recognizer: UITapGestureRecognizer) {
        log.verbose("User did tap for lock")
        
//        [self hideViewController:unlockedViewController withAppearanceTransition:YES force:YES];
//        [self showViewController:teaserViewController withAppearanceTransition:YES];
    }
    
}
